### PR TITLE
[slf4j]: Use SLF4J

### PR DESCRIPTION
**Pull request by sanmibuh**
Use SLF4J instead of java.util.logging #113

No more needed **silentMode** option. Logs have INFO level. If you do not want to log, simply configure your logging engine for not log **org.jeasy.rules** package at INFO level.

### CODE REVIEW

**Comment by benas** [comment]
Thank you for this PR!

The `silentMode` parameter should be removed in 2 steps: first it should be deprecated in next version and removed in the version after. Next version will be a minor version 3.1, we can't remove a public API directly.

Could you please remove it from this PR and open a separate issue/PR for that?

I've added a couple of notes regarding to your changes and then it should be ready to merge.

File: easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRuleListener.java [hunk 1]
```diff
@@ -56,11 +56,12 @@ public void beforeExecute(Rule rule, Facts facts) {
...
     public void onSuccess(Rule rule, Facts facts) {
-        LOGGER.log(Level.INFO, "Rule ''{0}'' performed successfully", rule.getName());
+        LOGGER.info("Rule ''{}'' performed successfully", rule.getName());
     }
 
     @Override
     public void onFailure(Rule rule, Facts facts, Exception exception) {
-        LOGGER.log(Level.SEVERE, String.format("Rule '%s' performed with error", rule.getName()), exception);
+        LOGGER.info("Rule '{}' performed with error: {}",
+            rule.getName(), exception.getMessage());
```
Comment (benas) [inline easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRuleListener.java]: The exception should not be swallowed here. We should log the whole exception, not only the message. Do you agree?

File: pom.xml [hunk 2]
```diff
@@ -85,6 +85,12 @@
             </dependency>
 
             <dependency>
+                <groupId>org.slf4j</groupId>
+                <artifactId>slf4j-api</artifactId>
+                <version>1.7.25</version>
```
Comment (benas) [inline pom.xml]: This could be extracted in a maven property

**Comment by sanmibuh** [comment]
I can restore silentMode parameter. But with SLF4J there is no way to deactivate programmatically loggers. Slf4j is a facade for a lot of logging engines, but not engine itself.
Restoring silentMode parameter will be only for compiling compatibility. It would do nothing.

Do you want to restore with this premise?

File: easy-rules-core/src/test/java/org/jeasy/rules/core/DefaultRulesEngineTest.java [hunk 3]
```diff
@@ -56,7 +58,6 @@
     public void setup() throws Exception {
         super.setup();
         when(rule1.getName()).thenReturn("r");
-        when(rule1.getDescription()).thenReturn("d");
```
Comment (benas) [inline easy-rules-core/src/test/java/org/jeasy/rules/core/DefaultRulesEngineTest.java]: I don't understand the removal of this. Is it related to the purpose of this PR?

File: easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRulesEngine.java [hunk 4]
```diff
@@ -217,29 +219,37 @@ private boolean shouldBeEvaluated(Rule rule, Facts facts) {
...
+        LOGGER.info("Skip on missing fact: {}",
+            parameters.isSkipOnMissingFact());
     }
 
     private void log(Rules rules) {
-        LOGGER.log(Level.INFO, "Registered rules:");
-        for (Rule rule : rules) {
-            LOGGER.log(Level.INFO, format("Rule { name = '%s', description = '%s', priority = '%s'}",
-                    rule.getName(), rule.getDescription(), rule.getPriority()));
+        if (LOGGER.isInfoEnabled()) {
```
Comment (benas) [inline easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRulesEngine.java]: Do we really need this test? If the level is info, the message will be logged. If the level is higher than info, the message will not be logged automatically. Or I'm missing something?

**Comment by benas** [comment]
Yes, this is for backward compatibility. users upgrading from 3.0 to 3.1 will be broken.
In semver semantics, no removal of public APIs in a minor version.

We can do it for v3.2.

File: easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRulesEngine.java (hunk 4 above)
Comment (sanmibuh) [reply easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRulesEngine.java]: This check avoids run the for. If the logger is not active, no reason to iterate.

File: easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRuleListener.java (hunk 1 above)
Comment (sanmibuh) [reply easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRuleListener.java]: Totally agree ;)

File: pom.xml (hunk 2 above)
Comment (sanmibuh) [reply pom.xml]: Totally agree

File: easy-rules-core/src/test/java/org/jeasy/rules/core/DefaultRulesEngineTest.java (hunk 3 above)
Comment (sanmibuh) [reply easy-rules-core/src/test/java/org/jeasy/rules/core/DefaultRulesEngineTest.java]: It was an IntelliJ warning. This mock isn't needed

File: easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRuleListener.java (hunk 1 above)
Comment (sanmibuh) [reply easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRuleListener.java]: Note that slf4j doesn't have natively a method with string replacement and exception logging. 
For this case we need manually concat the message (in this case checking first if is the log enabled avoiding concatenation when the log is off)

**Comment by sanmibuh** [comment]
Hello @benas 
You have the changes. When this PR is merged I´ll do another PR removing deprecated  _silentMode_   engine config. 👍 

Best

**Comment by benas** [comment]
Hello @sanmibuh 

Great! Thank you for these updates 👍 

I will merge your PR and include your changes in next release.
Keep tuned!

Kind regards
Mahmoud

**Comment by cemo** [comment]
@benas `info` level is default. I think that only an error case should be logged by default either by a `warn` or `error`. What I am suggesting is that using `debug` level might be a better idea? 

I expect default usage is less verbose. Current implementation is too verbose right now.

**Comment by benas** [comment]
@cemo I think most of the time, users would like to see these information by default (at least from my perspective). I am expecting to have these infos by default without any config. We can set the level to debug by default, but in this case, anyone new to easy rules would have to raise the level to info.. I prefer keeping info by default and let those who are not interested in these details raise the level to warn/error.

**Pull request merged**